File: Loenn/entities/outline_display.lua
local DrawableNinePatch = require("structs.drawable_nine_patch")
local DrawableSprite = require("structs.drawable_sprite")
local Depths = require("consts.object_depths")

local OutlineDisplay = {}
OutlineDisplay.name = "Aqua/Outline Display"
OutlineDisplay.depth = Depths.SolidsBelow
OutlineDisplay.fieldInformation = {
    color = {
        fieldType = "color",
        useAlpha = false,
    },
}
OutlineDisplay.placements = {
    name = "Outline Display",
    data = {
        width = 24,
        height = 24,
        texture_path = "objects/acceleration_area/area",
        color = "ffffff",
    },
}

local blockNinePatchOptions = {
    mode = "border",
    borderMode = "repeat"
}

function OutlineDisplay.sprite(room, entity)
    local ninePatch = DrawableNinePatch.fromTexture(entity.texture_path, blockNinePatchOptions, entity.x, entity.y, entity.width, entity.height)
    local sprites = {}
    for _, sprite in ipairs(ninePatch:getDrawableSprite()) do
        sprite:setColor(entity.color)
        table.insert(sprites, sprite)
    end
    return sprites
end

return OutlineDisplay

File: Source/Core/Entities/OutlineDisplay.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.Aqua.Core
{
    [CustomEntity("Aqua/Outline Display")]
    public class OutlineDisplay : Entity
    {
        public MTexture Texture { get; private set; }
        public Color Color { get; private set; }

        public OutlineDisplay(EntityData data, Vector2 offset)
            : base(data.Position + offset)
        {
            string path = data.Attr("texture_path", "objects/acceleration_area/area");
            if (!GFX.Game.Has(path))
            {
                path = "objects/acceleration_area/area";
            }
            Texture = GFX.Game[path];
            Color = data.HexColor("color", Color.White);
            var image = new Image9Slice(Texture, data.Width, data.Height, Image9Slice.RenderMode.Border);
            image.SetColor(Color);
            Add(image);
            Depth = Depths.SolidsBelow;
            Active = false;
        }
    }
}
